### PR TITLE
feat(dashboard): csv data hook

### PR DESCRIPTION
**Pull request by corteggiano**
## Overview
add useCSVData hook which pulls from the query and useDataStream hook to get data needed for a future CSV feature 

example graph (modeled data)
Screenshot 2023-10-31 at 15 29 48

example data (modeled data)
```
[
  {
    "id": "1a89034b-0705-416c-a67f-340e574affd8---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 43.26532776414541,
    "unit": "Degrees",
    "timestamp": "2023/10/28 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "1a89034b-0705-416c-a67f-340e574affd8",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 4"
  },
  {
    "id": "1a89034b-0705-416c-a67f-340e574affd8---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 51.44607894935983,
    "unit": "Degrees",
    "timestamp": "2023/10/29 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "1a89034b-0705-416c-a67f-340e574affd8",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 4"
  },
  {
    "id": "1a89034b-0705-416c-a67f-340e574affd8---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 36.235690329124985,
    "unit": "Degrees",
    "timestamp": "2023/10/30 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "1a89034b-0705-416c-a67f-340e574affd8",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 4"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "value": 23.083161628791185,
    "unit": "m/s",
    "timestamp": "2023/10/28 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "dataType": "DOUBLE",
    "propertyName": "Wind Speed",
    "assetName": "Demo Turbine Asset 2"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "value": 21.41656834542248,
    "unit": "m/s",
    "timestamp": "2023/10/29 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "dataType": "DOUBLE",
    "propertyName": "Wind Speed",
    "assetName": "Demo Turbine Asset 2"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "value": 30.83296248057125,
    "unit": "m/s",
    "timestamp": "2023/10/30 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "b956e05b-a04f-4b42-966a-e76b22f0f5ac",
    "dataType": "DOUBLE",
    "propertyName": "Wind Speed",
    "assetName": "Demo Turbine Asset 2"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 46.87695050390162,
    "unit": "Degrees",
    "timestamp": "2023/10/28 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 2"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 45.165212789575904,
    "unit": "Degrees",
    "timestamp": "2023/10/29 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 2"
  },
  {
    "id": "bdae7633-3c51-4d1e-a463-e6a2c098afa1---c4b0344e-4961-4e16-854d-05642ffd96c1",
    "value": 33.78083931963757,
    "unit": "Degrees",
    "timestamp": "2023/10/30 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000,
    "assetId": "bdae7633-3c51-4d1e-a463-e6a2c098afa1",
    "propertyId": "c4b0344e-4961-4e16-854d-05642ffd96c1",
    "dataType": "DOUBLE",
    "propertyName": "Wind Direction",
    "assetName": "Demo Turbine Asset 2"
  }
]
```

example graph (unmodeled data)
Screenshot 2023-10-31 at 15 31 03

example data (unmodeled data)
```
[
  {
    "id": "DemoDisassocaitedAlias",
    "value": 3.4364159754343238,
    "timestamp": "2023/10/28 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000
  },
  {
    "id": "DemoDisassocaitedAlias",
    "value": 3.181893748346945,
    "timestamp": "2023/10/29 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000
  },
  {
    "id": "DemoDisassocaitedAlias",
    "value": 2.4758414402321436,
    "timestamp": "2023/10/30 20:00:00",
    "aggregationType": "AVERAGE",
    "resolution": 86400000
  }
]
```

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: packages/dashboard/src/components/csvDownloadButton/useCSVData.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    // absolute time range
+    const duration = parseDuration(viewport.duration);
+    return currentPoint >= new Date(timeOfRequestMS - duration) && currentPoint <= new Date(timeOfRequestMS);
+  } else {
+    // relative time range
+    return currentPoint >= viewport.start && currentPoint <= viewport.end;
+  }
+};
+
+export const useCSVData = ({
```
Comment (chejimmy) [inline packages/dashboard/src/components/csvDownloadButton/useCSVData.ts:31]: nit: rename to `useViewportData`? not exactly CSV data but flatten data values during the viewport;
is the plan to wrap this hook under another hook for parsing and downloding?

**Review by chejimmy — APPROVED** [review]
LGTM!

**Review by dpportet — COMMENTED** [review]
There's some linting errors to fix for the unit tests to pass.

File: packages/dashboard/src/components/csvDownloadButton/types.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+export type CSVDownloadObject = {
+  id: string;
+  value: any;
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/dashboard/src/components/csvDownloadButton/types.ts:3]: yeah, should be Primitive!

**Pull request merged**